libff/algebra/fields/field_utils.tcc

```diff
@@ -74,8 +85,83 @@ size_t field_get_digit(
+    const size_t low_val_shifted = low_val >> low_limb_shift;
+    const size_t low_val_final = use_low_limb * low_val_shifted;
+
+    const size_t high_val = v.data[use_high_limb * high_limb];
+    const size_t high_val_shifted = high_val << high_limb_shift;
+    const size_t high_val_final = use_high_limb * high_val_shifted;
+
+    const size_t value = mask & (low_val_final | high_val_final);
+
+    assert((low_limb < n) || (value == 0));
+    return value;
+}
+
+template<typename FieldT>
+size_t field_get_num_signed_digits(const size_t digit_size)
+{
+    // -1 in the field must have the largest number of 1's in the higher order
+    // elements of its binary representation.
+    const bigint<FieldT::num_limbs> minus_one = (-FieldT::one()).as_bigint();
+
+    // The naive approach is to compute:
+    //
+    //   ciel(FieldT::num_bits + 1 / digit_size)
+    //
+    // where the extra bit avoids the need to overflow in the final digit.
+    // However there are edges cases. Consider digit size 2 for a field where
+    // -1 has bits: 1 1 1 0 0 ....  with 2-bit digit boundaries:
+    //
+    //    1 | 11 | 00
+    //
+    // The extra bit added in the calculation above becomes the SIGN bit of the
+    // high-order digit:
+    //
+    //   01 | 11 | 00
+    //
+    // and the second digit is guaranteed to overflow, in turn causing the 1st
+    // digit to overflow. We wish to minimize the number of digits used and
+    // rely on the naive calculation where possible. Therefore the actual value
+    // of the digits of -1 must be examined.
+
+    // Examine the unsigned digits, checking for the case where the final digit
+    // overflows.
+
+    const size_t naive_num_digits =
+        (FieldT::num_bits + 1 + digit_size - 1) / digit_size;
+    const size_t sign_bit_mask = 1ull << (digit_size - 1);
+    const size_t max_signed_value = sign_bit_mask - 1;
+
+    // Check each unsigned digit, starting with the highest order. If it cannot
+    // overflow (< max_signed_value) then we can early-out and use
+    // naive_num_digits. If it will definitely overflow, we need an extra
+    // digit. If it is max_signed_value, move to the next least significant
+    // digit and repeat the check.
+
+    bool final_overflow = false;
+    for (size_t i = naive_num_digits - 1; i < naive_num_digits; --i) {
+        const size_t unsigned_digit = field_get_digit(minus_one, digit_size, i);
+        if (unsigned_digit & sign_bit_mask) {
+            final_overflow = true;
+            break;
+        }
+
+        if (unsigned_digit != max_signed_value) {
+            break;
+        }
+
+        // This (and any previous values) are max_signed_value. Any lower-order
+        // digits which have the signed bit set will trigger an overflow up to
+        // the most-significant digit.
+    }
+
+    if (!final_overflow) {
+        return naive_num_digits;
+    } else {
+        return naive_num_digits + 1;
+    }
```

Comment:
Why not remove the `else` part here and simply do:

```cpp
    if (!final_overflow) {
        return naive_num_digits;
    }
    
    return naive_num_digits + 1;
```
?

Or the other way around
```cpp
if (final_overflow) {
    return naive_num_digits + 1;
}
return naive_num_digits;
```

Reply:
Removed the `else`.